src/xsdata.cpp

```diff
@@ -0,0 +1,700 @@
+    if (scatter_format == ANGLE_HISTOGRAM) {
+      scatter[a] = new ScattDataHistogram;
+    } else if (scatter_format == ANGLE_TABULAR) {
+      scatter[a] = new ScattDataTabular;
+    } else if (scatter_format == ANGLE_LEGENDRE) {
+      scatter[a] = new ScattDataLegendre;
+    }
+  }
+}
+
+//==============================================================================
+
+void
+XsData::from_hdf5(hid_t xsdata_grp, bool fissionable, int scatter_format,
+     int final_scatter_format, int order_data, int max_order,
+     int legendre_to_tabular_points, bool is_isotropic, int n_pol, int n_azi)
+{
+  // Reconstruct the dimension information so it doesn't need to be passed
+  int n_ang = n_pol * n_azi;
+  int energy_groups = total[0].size();
+  int delayed_groups = decay_rate[0].size();
+
+  // Set the fissionable-specific data
+  if (fissionable) {
+    fission_from_hdf5(xsdata_grp, n_pol, n_azi, energy_groups, delayed_groups,
+                      is_isotropic);
+  }
+  // Get the non-fission-specific data
+  read_nd_vector(xsdata_grp, "decay_rate", decay_rate);
+  read_nd_vector(xsdata_grp, "absorption", absorption, true);
+  read_nd_vector(xsdata_grp, "inverse-velocity", inverse_velocity);
+
+  // Get scattering data
+  scatter_from_hdf5(xsdata_grp, n_pol, n_azi, energy_groups, scatter_format,
+       final_scatter_format, order_data, max_order, legendre_to_tabular_points);
+
+  // Check absorption to ensure it is not 0 since it is often the
+  // denominator in tally methods
+  for (int a = 0; a < n_ang; a++) {
+    for (int gin = 0; gin < energy_groups; gin++) {
+      if (absorption[a][gin] == 0.) absorption[a][gin] = 1.e-10;
+    }
+  }
+
+  // Get or calculate the total x/s
+  if (object_exists(xsdata_grp, "total")) {
+    read_nd_vector(xsdata_grp, "total", total);
+  } else {
+    for (int a = 0; a < n_ang; a++) {
+      for (int gin = 0; gin < energy_groups; gin++) {
+        total[a][gin] = absorption[a][gin] + scatter[a]->scattxs[gin];
+      }
+    }
+  }
+
+  // Fix if total is 0, since it is in the denominator when tallying
+  for (int a = 0; a < n_ang; a++) {
+    for (int gin = 0; gin < energy_groups; gin++) {
+      if (total[a][gin] == 0.) total[a][gin] = 1.e-10;
+    }
+  }
+}
+
+//==============================================================================
+
+void
+XsData::fission_from_hdf5(hid_t xsdata_grp, int n_pol, int n_azi,
+     int energy_groups, int delayed_groups, bool is_isotropic)
+{
+  int n_ang = n_pol * n_azi;
+  // Get the fission and kappa_fission data xs; these are optional
+  read_nd_vector(xsdata_grp, "fission", fission);
+  read_nd_vector(xsdata_grp, "kappa-fission", kappa_fission);
+
+  // Set/get beta
+  double_3dvec temp_beta =double_3dvec(n_ang, double_2dvec(energy_groups,
+       double_1dvec(delayed_groups, 0.)));
+  if (object_exists(xsdata_grp, "beta")) {
+    hid_t xsdata = open_dataset(xsdata_grp, "beta");
+    int ndims = dataset_ndims(xsdata);
+
+    if (is_isotropic) ndims += 2;
```

Comment:
Is it not the other way around? Shouldn't there be more dimensions if it's not isotropic?

Reply:
This is an artifact of the hdf5 library having different dimensions in the data if it is isotropic or angular but the code here making only a minimal distinction between the two.  So this says "if the library is isotropic, the number of dimensions you will see will be 2 too low for the comparisons we are about to make".